File: go/divide_and_conquer/108_sorted_array_to_bst.go
package divideandconquer

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// sortedArrayToBST 将有序数组转换为二叉搜索树
func sortedArrayToBST(nums []int) *TreeNode {

	return helper(nums, 0, len(nums)-1)
}

// 递归，总是选择中间元素 靠左的座位跟节点
func helper(nums []int, left, right int) *TreeNode {

	if left > right {
		return nil
	}

	mid := left + (right-left)>>1
	root := &TreeNode{Val: nums[mid]}

	root.Left = helper(nums, left, mid-1)
	root.Right = helper(nums, mid+1, right)

	return root
}

File: rust/src/divide_and_conquer/_108_sorted_array_to_bst.rs
use super::TreeNode;
use std::cell::RefCell;
use std::rc::Rc;

pub fn sorted_array_to_bst(nums: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
    helper(&nums, 0, nums.len() as i32 - 1)
}

fn helper(nums: &Vec<i32>, left: i32, right: i32) -> Option<Rc<RefCell<TreeNode>>> {
    if left > right {
        return None;
    }

    let mid = left + ((right - left) >> 1);

    let root = Rc::new(RefCell::new(TreeNode::new(mid)));

    root.borrow_mut().left = helper(nums, left, mid - 1);
    root.borrow_mut().right = helper(nums, mid + 1, right);

    Some(root)
}
